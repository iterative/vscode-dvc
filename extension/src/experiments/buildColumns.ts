import {
  ExperimentsRepoJSONOutput,
  ValueTree,
  Value,
  DataDictRoot,
  ExperimentJSONOutput
} from './contract'

interface BuildColumnsOutput {
  params?: Column[]
  metrics?: Column[]
}

interface InferredColumns {
  nestedColumns: Column[]
}

interface ColumnCommon {
  maxStringLength?: number
  maxNumber?: number
  minNumber?: number
}

interface PartialColumnDescriptor extends ColumnCommon {
  types?: Set<string>
  childColumns?: PartialColumnsMap
}
type PartialColumnsMap = Map<string, PartialColumnDescriptor>

export interface Column extends ColumnCommon {
  name: string
  types?: string[]
  childColumns?: Column[]
  ancestors?: string[]
}

interface AggregatedColumns {
  paramsMap: PartialColumnsMap | undefined
  metricsMap: PartialColumnsMap | undefined
}

const getValueType = (value: Value | ValueTree) => {
  if (value === null) {
    return 'null'
  }
  return typeof value
}

const mergeNumberColumn = (
  columnDescriptor: PartialColumnDescriptor,
  newNumber: number
): void => {
  const { maxNumber, minNumber } = columnDescriptor
  if (maxNumber === undefined || maxNumber < newNumber) {
    columnDescriptor.maxNumber = newNumber
  }
  if (minNumber === undefined || minNumber > newNumber) {
    columnDescriptor.minNumber = newNumber
  }
}

const mergePrimitiveColumn = (
  columnDescriptor: PartialColumnDescriptor,
  newValue: Value,
  newValueType: string
): PartialColumnDescriptor => {
  const { maxStringLength } = columnDescriptor

  const stringifiedAddition = String(newValue)
  const additionStringLength = stringifiedAddition.length
  if (maxStringLength === undefined || maxStringLength < additionStringLength) {
    columnDescriptor.maxStringLength = additionStringLength
  }

  if (newValueType === 'number') {
    mergeNumberColumn(columnDescriptor, newValue as number)
  }

  return columnDescriptor as PartialColumnDescriptor
}

const mergeOrCreateColumnsMap = (
  originalColumnsMap: PartialColumnsMap = new Map(),
  valueTree: ValueTree | DataDictRoot
): PartialColumnsMap => {
  const sampleEntries = Object.entries(valueTree)
  for (const [propertyKey, propertyValue] of sampleEntries) {
    originalColumnsMap.set(
      propertyKey,
      // eslint-disable-next-line @typescript-eslint/no-use-before-define
      mergeOrCreateColumnDescriptor(
        originalColumnsMap.get(propertyKey),
        propertyValue
      )
    )
  }
  return originalColumnsMap
}

const mergeOrCreateColumnDescriptor = (
  columnDescriptor: PartialColumnDescriptor = {},
  newValue: Value | ValueTree
): PartialColumnDescriptor => {
  const newValueType = getValueType(newValue)

  if (newValueType === 'object') {
    columnDescriptor.childColumns = mergeOrCreateColumnsMap(
      columnDescriptor.childColumns,
      newValue as ValueTree
    )
    return columnDescriptor as PartialColumnDescriptor
  } else {
    if (!columnDescriptor.types) {
      columnDescriptor.types = new Set()
    }
    const { types } = columnDescriptor
    types.add(newValueType)
    return mergePrimitiveColumn(
      columnDescriptor,
      newValue as Value,
      newValueType
    )
  }
}

const columnFromMapEntry = (
  entry: [string, PartialColumnDescriptor]
): Column => {
  const [name, partialColumnDescriptor] = entry
  const { types, maxStringLength, minNumber, maxNumber } =
    partialColumnDescriptor
  const column: Column = {
    name
  }
  if (maxStringLength) {
    column.maxStringLength = maxStringLength
  }
  if (minNumber) {
    column.minNumber = minNumber
    column.maxNumber = maxNumber
  }
  if (types) {
    column.types = [...types]
  }
  return column
}

const transformAndCollectFromColumns = (
  columnsMap: PartialColumnsMap,
  ancestors?: string[]
): Column[] => {
  const currentLevelColumns = []
  for (const entry of columnsMap) {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    currentLevelColumns.push(buildColumn(entry, ancestors))
  }
  return currentLevelColumns
}

const transformColumnsMap = (
  columnsMap: PartialColumnsMap
): InferredColumns => {
  const nestedColumns = transformAndCollectFromColumns(columnsMap)
  return { nestedColumns }
}

const buildColumn = (
  entry: [string, PartialColumnDescriptor],
  ancestors?: string[]
): Column => {
  const finalColumn = columnFromMapEntry(entry)

  const [name, { childColumns }] = entry

  if (ancestors) {
    finalColumn.ancestors = ancestors
  }

  if (childColumns) {
    finalColumn.childColumns = transformAndCollectFromColumns(
      childColumns,
      ancestors ? [...ancestors, name] : [name]
    )
  }

  return finalColumn
}

const walkExperiments = (
  tableData: ExperimentsRepoJSONOutput,
  onExperiment: (commit: ExperimentJSONOutput) => void
) => {
  for (const branch of Object.values(tableData)) {
    for (const commit of Object.values(branch)) {
      onExperiment(commit)
    }
  }
}

const aggregateColumns = (
  tableData: ExperimentsRepoJSONOutput
): AggregatedColumns => {
  let paramsMap: PartialColumnsMap | undefined
  let metricsMap: PartialColumnsMap | undefined

  walkExperiments(tableData, (commit: ExperimentJSONOutput) => {
    const { params, metrics } = commit
    if (params) {
      paramsMap = mergeOrCreateColumnsMap(paramsMap, params)
    }
    if (metrics) {
      metricsMap = mergeOrCreateColumnsMap(metricsMap, metrics)
    }
  })

  return { metricsMap, paramsMap }
}

const buildColumnsOutput = ({
  paramsMap,
  metricsMap
}: AggregatedColumns): BuildColumnsOutput => {
  const output: BuildColumnsOutput = {}

  if (paramsMap) {
    const { nestedColumns } = transformColumnsMap(paramsMap)
    output.params = nestedColumns
  }

  if (metricsMap) {
    const { nestedColumns } = transformColumnsMap(metricsMap)
    output.metrics = nestedColumns
  }

  return output
}

export const buildColumns = (
  tableData: ExperimentsRepoJSONOutput
): BuildColumnsOutput => {
  return buildColumnsOutput(aggregateColumns(tableData))
}
