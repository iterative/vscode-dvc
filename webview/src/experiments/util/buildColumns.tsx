import React from 'react'
import get from 'lodash/get'
import {
  createColumnHelper,
  AccessorFn,
  Column as TableColumn,
  ColumnDef,
  CellContext
} from '@tanstack/react-table'
import {
  Column,
  ColumnType,
  Experiment
} from 'dvc/src/experiments/webview/contract'
import { Header } from '../components/table/content/Header'
import { Cell, CellValue } from '../components/table/content/Cell'
import { TimestampHeader } from '../components/table/content/TimestampHeader'
import { DateCellContents } from '../components/table/content/DateCellContent'

export type ColumnWithGroup = ColumnDef<Experiment, unknown> & {
  group: ColumnType
}

export const columnHelper = createColumnHelper<Column>()

const buildAccessor: (valuePath: string[]) => AccessorFn<Column> =
  pathArray => originalRow => {
    const value = get(originalRow, pathArray)

    if (!Array.isArray(value)) {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return
      return value
    }
    return `[${value.join(', ')}]`
  }

const getMainColumnProperties = (
  type: ColumnType,
  label: string,
  path: string,
  size?: number
) => {
  const basicProperties = {
    group: type
  }
  const sizeProperty = size ? { size } : {}

  if (type === ColumnType.TIMESTAMP) {
    return {
      ...basicProperties,
      ...sizeProperty,
      cell: DateCellContents as unknown as React.FC<
        CellContext<Column, CellValue>
      >,
      header: () => <TimestampHeader />,
      id: 'Created'
    }
  }

  return {
    ...basicProperties,
    ...sizeProperty,
    cell: Cell as unknown as React.FC<CellContext<Column, CellValue>>,
    header: () => <Header name={label} />,
    id: path
  }
}

export const buildColumns = (
  properties: Column[],
  parentPath: string
): TableColumn<Experiment>[] => {
  return properties
    .filter(column => column.parentPath === parentPath)
    .map(data => {
      const { path, width, pathArray, label, type } = data

      const childColumns = buildColumns(properties, path)
      const mainColumnProperties = getMainColumnProperties(
        type,
        label,
        path,
        width
      )

      if (childColumns.length > 0) {
        return columnHelper.group({
          ...mainColumnProperties,
          cell: undefined,
          columns: childColumns
        })
      }

      return columnHelper.accessor(buildAccessor(pathArray || [path]), {
        ...mainColumnProperties
      })
    })
    .filter(Boolean) as TableColumn<Experiment>[]
}
