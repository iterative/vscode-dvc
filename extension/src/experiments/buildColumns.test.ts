import { buildColumns, Column } from './buildColumns'

describe('buildColumns', () => {
  it('Outputs both params and metrics when both are present', () => {
    const { params, metrics } = buildColumns({
      workspace: {
        baseline: {
          metrics: {
            1: {
              2: 3
            }
          },
          params: {
            a: {
              b: 'c'
            }
          }
        }
      }
    })
    expect(params).toBeDefined()
    expect(metrics).toBeDefined()
  })

  it('Omits params when none exist in the source data', () => {
    const { params, metrics } = buildColumns({
      workspace: {
        baseline: {
          metrics: {
            1: {
              2: 3
            }
          }
        }
      }
    })
    expect(params).toBeUndefined()
    expect(metrics).toBeDefined()
  })

  it('returns an empty object if input data is empty', () => {
    const output = buildColumns({
      workspace: {
        baseline: {}
      }
    })
    expect(output).toEqual({})
  })

  describe('minimal mixed column example', () => {
    const exampleBigNumber = 3000000000
    const { params } = buildColumns({
      brancha: {
        baseline: {
          params: {
            'params.yaml': {
              mixedparam: 'string'
            }
          }
        },
        otherexp: {
          params: {
            'params.yaml': {
              mixedparam: true
            }
          }
        }
      },
      branchb: {
        baseline: {
          params: {
            'params.yaml': {
              mixedparam: null
            }
          }
        }
      },
      workspace: {
        baseline: {
          params: {
            'params.yaml': {
              mixedparam: exampleBigNumber
            }
          }
        }
      }
    }) as {
      params: Column[]
    }
    const [paramsFileColumn] = params
    const [exampleMixedColumn] = paramsFileColumn.childColumns as Column[]

    it('correctly identifies mixed type params', () =>
      expect(exampleMixedColumn.types).toEqual([
        'string',
        'boolean',
        'null',
        'number'
      ]))

    it('correctly identifies a number as the highest string length of a mixed column', () =>
      expect(exampleMixedColumn.maxStringLength).toEqual(10))

    it('adds a highest and lowest number from the one present', () => {
      expect(exampleMixedColumn.maxNumber).toEqual(exampleBigNumber)
      expect(exampleMixedColumn.minNumber).toEqual(exampleBigNumber)
    })
  })

  it('finds different minNumber and maxNumber on a mixed column', () => {
    const { params } = buildColumns({
      workspace: {
        baseline: {
          params: {
            'params.yaml': {
              mixednumber: null
            }
          }
        },
        exp1: {
          params: {
            'params.yaml': {
              mixednumber: 0
            }
          }
        },
        exp2: {
          params: {
            'params.yaml': {
              mixednumber: -1
            }
          }
        },
        exp3: {
          params: {
            'params.yaml': {
              mixednumber: 1
            }
          }
        }
      }
    }) as { params: Column[] }
    const [paramsFileColumn] = params
    const [mixedColumn] = paramsFileColumn.childColumns as Column[]

    expect(mixedColumn.minNumber).toEqual(-1)
    expect(mixedColumn.maxNumber).toEqual(1)
  })

  describe('Number features', () => {
    const { params } = buildColumns({
      workspace: {
        baseline: {
          params: {
            'params.yaml': {
              withNumbers: -1,
              withoutNumbers: 'a'
            }
          }
        },
        exp1: {
          params: {
            'params.yaml': {
              withNumbers: 2,
              withoutNumbers: 'b'
            }
          }
        },
        exp2: {
          params: {
            'params.yaml': {
              withNumbers: 'c',
              withoutNumbers: 'b'
            }
          }
        }
      }
    }) as {
      params: Column[]
    }
    const [paramsFileColumn] = params
    const [columnWithNumbers, columnWithoutNumbers] =
      paramsFileColumn.childColumns as Column[]

    it('does not add maxNumber or minNumber on a column with no numbers', () => {
      expect(columnWithoutNumbers.minNumber).toBeUndefined()
      expect(columnWithoutNumbers.maxNumber).toBeUndefined()
    })
    it('finds the min number of -1', () =>
      expect(columnWithNumbers.minNumber).toEqual(-1))
    it('finds the max number of 2', () =>
      expect(columnWithNumbers.maxNumber).toEqual(2))
    it('finds a max string length of two from -1', () =>
      expect(columnWithNumbers.maxStringLength).toEqual(2))
  })

  it('aggregates multiple different field names', () => {
    const { params } = buildColumns({
      brancha: {
        baseline: {
          params: {
            'params.yaml': {
              one: 1
            }
          }
        },
        otherexp: {
          params: {
            'params.yaml': {
              two: 2
            }
          }
        }
      },
      branchb: {
        baseline: {
          params: {
            'params.yaml': {
              three: 3
            }
          }
        }
      },
      workspace: {
        baseline: {
          params: {
            'params.yaml': {
              four: 4
            }
          }
        }
      }
    })

    const [paramsYamlColumn] = params as Column[]
    const paramsColumns = paramsYamlColumn.childColumns as Column[]

    expect(paramsColumns?.map(({ name }) => name)).toEqual([
      'one',
      'two',
      'three',
      'four'
    ])
  })

  it('does not report types for columns without primitives or children for columns without objects', () => {
    const { params } = buildColumns({
      workspace: {
        baseline: {
          params: {
            'params.yaml': {
              onlyHasChild: {
                onlyHasPrimitive: 1
              }
            }
          }
        }
      }
    })

    const [paramsYamlColumn] = params as Column[]
    const [objectColumn] = paramsYamlColumn.childColumns as Column[]

    expect(objectColumn.name).toEqual('onlyHasChild')
    expect(objectColumn.childColumns).toBeDefined()
    expect(objectColumn.types).toBeUndefined()

    const primitiveColumn = (objectColumn.childColumns as Column[])[0]

    expect(primitiveColumn.name).toEqual('onlyHasPrimitive')
    expect(primitiveColumn.types).toBeDefined()
    expect(primitiveColumn.childColumns).toBeUndefined()
  })
})
