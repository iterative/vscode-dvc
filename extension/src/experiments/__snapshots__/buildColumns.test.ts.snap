// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildColumns parses minimal exp show data artificially stuffed with corner cases 1`] = `
Array [
  Array [
    Object {
      "childColumns": Array [
        Object {
          "ancestors": Array [
            "params",
          ],
          "childColumns": Array [
            Object {
              "ancestors": Array [
                "params",
                "params.yaml",
              ],
              "maxStringLength": 6,
              "name": "mixedparam",
              "types": Array [
                "number",
                "string",
              ],
            },
            Object {
              "ancestors": Array [
                "params",
                "params.yaml",
              ],
              "maxStringLength": 5,
              "name": "boolparam",
              "types": Array [
                "boolean",
              ],
            },
            Object {
              "ancestors": Array [
                "params",
                "params.yaml",
              ],
              "maxStringLength": 6,
              "name": "stringparam",
              "types": Array [
                "string",
              ],
            },
            Object {
              "ancestors": Array [
                "params",
                "params.yaml",
              ],
              "maxStringLength": 9,
              "name": "undefparam",
              "types": Array [
                "undefined",
              ],
            },
          ],
          "name": "params.yaml",
        },
      ],
      "name": "params",
    },
    Object {
      "childColumns": Array [
        Object {
          "ancestors": Array [
            "metrics",
          ],
          "childColumns": Array [
            Object {
              "ancestors": Array [
                "metrics",
                "logs.json",
              ],
              "maxStringLength": 6,
              "name": "acc",
              "types": Array [
                "number",
              ],
            },
            Object {
              "ancestors": Array [
                "metrics",
                "logs.json",
              ],
              "maxStringLength": 18,
              "name": "loss",
              "types": Array [
                "number",
              ],
            },
            Object {
              "ancestors": Array [
                "metrics",
                "logs.json",
              ],
              "maxStringLength": 1,
              "name": "step",
              "types": Array [
                "number",
              ],
            },
            Object {
              "ancestors": Array [
                "metrics",
                "logs.json",
              ],
              "maxStringLength": 4,
              "name": "boolparam",
              "types": Array [
                "boolean",
              ],
            },
            Object {
              "ancestors": Array [
                "metrics",
                "logs.json",
              ],
              "maxStringLength": 4,
              "name": "nullparam",
              "types": Array [
                "null",
              ],
            },
          ],
          "name": "logs.json",
        },
      ],
      "name": "metrics",
    },
  ],
  Array [
    Object {
      "ancestors": Array [
        "params",
        "params.yaml",
      ],
      "maxStringLength": 6,
      "name": "mixedparam",
      "types": Array [
        "number",
        "string",
      ],
    },
    Object {
      "ancestors": Array [
        "params",
        "params.yaml",
      ],
      "maxStringLength": 5,
      "name": "boolparam",
      "types": Array [
        "boolean",
      ],
    },
    Object {
      "ancestors": Array [
        "params",
        "params.yaml",
      ],
      "maxStringLength": 6,
      "name": "stringparam",
      "types": Array [
        "string",
      ],
    },
    Object {
      "ancestors": Array [
        "params",
        "params.yaml",
      ],
      "maxStringLength": 9,
      "name": "undefparam",
      "types": Array [
        "undefined",
      ],
    },
    Object {
      "ancestors": Array [
        "metrics",
        "logs.json",
      ],
      "maxStringLength": 6,
      "name": "acc",
      "types": Array [
        "number",
      ],
    },
    Object {
      "ancestors": Array [
        "metrics",
        "logs.json",
      ],
      "maxStringLength": 18,
      "name": "loss",
      "types": Array [
        "number",
      ],
    },
    Object {
      "ancestors": Array [
        "metrics",
        "logs.json",
      ],
      "maxStringLength": 1,
      "name": "step",
      "types": Array [
        "number",
      ],
    },
    Object {
      "ancestors": Array [
        "metrics",
        "logs.json",
      ],
      "maxStringLength": 4,
      "name": "boolparam",
      "types": Array [
        "boolean",
      ],
    },
    Object {
      "ancestors": Array [
        "metrics",
        "logs.json",
      ],
      "maxStringLength": 4,
      "name": "nullparam",
      "types": Array [
        "null",
      ],
    },
  ],
]
`;

exports[`buildColumns parses real exp show data 1`] = `
Array [
  Array [
    Object {
      "childColumns": Array [
        Object {
          "ancestors": Array [
            "params",
          ],
          "childColumns": Array [
            Object {
              "ancestors": Array [
                "params",
                "params.yaml",
              ],
              "maxStringLength": 6,
              "name": "lr",
              "types": Array [
                "number",
              ],
            },
            Object {
              "ancestors": Array [
                "params",
                "params.yaml",
              ],
              "maxStringLength": 6,
              "name": "seed",
              "types": Array [
                "number",
              ],
            },
            Object {
              "ancestors": Array [
                "params",
                "params.yaml",
              ],
              "maxStringLength": 1,
              "name": "weight_decay",
              "types": Array [
                "number",
              ],
            },
          ],
          "name": "params.yaml",
        },
      ],
      "name": "params",
    },
    Object {
      "childColumns": Array [
        Object {
          "ancestors": Array [
            "metrics",
          ],
          "childColumns": Array [
            Object {
              "ancestors": Array [
                "metrics",
                "logs.json",
              ],
              "maxStringLength": 6,
              "name": "acc",
              "types": Array [
                "number",
              ],
            },
            Object {
              "ancestors": Array [
                "metrics",
                "logs.json",
              ],
              "maxStringLength": 18,
              "name": "loss",
              "types": Array [
                "number",
              ],
            },
            Object {
              "ancestors": Array [
                "metrics",
                "logs.json",
              ],
              "maxStringLength": 1,
              "name": "step",
              "types": Array [
                "number",
              ],
            },
          ],
          "name": "logs.json",
        },
      ],
      "name": "metrics",
    },
  ],
  Array [
    Object {
      "ancestors": Array [
        "params",
        "params.yaml",
      ],
      "maxStringLength": 6,
      "name": "lr",
      "types": Array [
        "number",
      ],
    },
    Object {
      "ancestors": Array [
        "params",
        "params.yaml",
      ],
      "maxStringLength": 6,
      "name": "seed",
      "types": Array [
        "number",
      ],
    },
    Object {
      "ancestors": Array [
        "params",
        "params.yaml",
      ],
      "maxStringLength": 1,
      "name": "weight_decay",
      "types": Array [
        "number",
      ],
    },
    Object {
      "ancestors": Array [
        "metrics",
        "logs.json",
      ],
      "maxStringLength": 6,
      "name": "acc",
      "types": Array [
        "number",
      ],
    },
    Object {
      "ancestors": Array [
        "metrics",
        "logs.json",
      ],
      "maxStringLength": 18,
      "name": "loss",
      "types": Array [
        "number",
      ],
    },
    Object {
      "ancestors": Array [
        "metrics",
        "logs.json",
      ],
      "maxStringLength": 1,
      "name": "step",
      "types": Array [
        "number",
      ],
    },
  ],
]
`;
